### ISSUE TITLE
How can we optimize the ring buffer performance ?

### ISSUE THREAD

**Issue by codehobbyist06** [NULLx76/ringbuffer#123]
Hi everyone,
I have been trying to compare performance of dpdk rings to AllocRingbuffer and have noticed that the results are significantly different,

Following are some stats:
buffer             No. of cycles to enqueue      No. of cycles to dequeue
dpdk ring                     4                                               6
rust ring                       41                                             68

So, as you can see there is difference of approximately 10 times, which makes it significantly slower. Hence, I wanted to know, if there is any way to optimize the ring buffer performance further, by using some flags etc. ?
Also, one difference the AllocRingBuffer has with the dpdk ring is that it does not have any bulk enqueue or dequeue APIs. So, is there any plan of having such APIs implemented?

**Comment by jdonszelmann** [comment]
Hi!

I am curious how you measured these statistics. not at all because I don't believe them, I'm certain there are ways to still optimise ringbuffer. Did you use our own benchmark and extrapola

About bulk insert apis: possibly! We're developing ringbuffer extremely part time so it's something I'd love to do if I find some time but it could take another month or so and I can't promise much. If you'd like to make an issue for it, feel free to.

**Comment by codehobbyist06** [comment]
I have measured the statistics by simply fetching the time register value before and after the operations. Also, both the rings have been compared in the same environment so I doubt, something could be wrong there. 
And regarding flags, actually I am relatively a bit new to rust so have not tried using any optimization flags as such. However, would like to know if I can use any such flags for better performance.
Also, the enqueue and deque operations are happening in different dpdk threads for the benchmarking, so that both the rings are on the same page.
I would really appreciate your inputs on how the ring performance could be optimized.

**Comment by jdonszelmann** [comment]
I see, well the first thing you can try is to compile with `--release` if you didn't do that yet. That might make a large difference if you didn't do that yet

**Comment by codehobbyist06** [comment]
Yes, I had compiled with --release flag for benchmarking. Also, apart from this, I did not use any optimization flags on rust side. The optimization setting are mostly the ones provided by default.

**Comment by jdonszelmann** [comment]
--release will optimise with optimisation level 3 so that's good. Another option you can try is enabling LTO

**Comment by codehobbyist06** [comment]
Ohk sure. I will try with that. Also, is there any performance optimization available from the ring side. For e.g : some ring configurations that could be used?

**Comment by jdonszelmann** [comment]
There is not any configuration you can pass to a ringbuffer to make it faster. ConstGenericRingBuffer is a bit faster than AllocRingBuffer but it depends on your usecase if that's useful to you. One more thing to know if that RingBuffer stores full elements, not references to elements. If your elements are large, they're copied around into and out of the RingBuffer. You could have a ringbuffer of 'static references (or whichever lifetime available to you) and your performance may change.

**Comment by codehobbyist06** [comment]
Ok sure. I will check the ConstGenericRingBuffer once if that could be useful. Thanks a lot for the info:)

Also, I wanted to keep static references in the ring buffer, but since the ring needs to have ownership of the objects (its a requirement on my side) it contains, I don't think I can do much on that part. However, would be open to know if there is some way to transfer the ownership of the objects to the ring as well as avoid copying the complete data chunk back and forth.

**Comment by jdonszelmann** [comment]
How do you do that in your C version? If I saw it correctly, it also mainly stores pointers or not?

**Comment by jdonszelmann** [comment]
With Box you can allocate first (which is expensive) and then pass only references around. If you're passing around the same references over and over again that may be worth it. You can also allocate in other places. Not on the heap, but maybe in an arena/bump allocator

**Comment by jdonszelmann** [comment]
If your references are really static, ringbuffer doesn't *need* ownership. It just depends on the generic type you're using. If you make a `RingBuffer::<&'static T>` RingBuffer doesn't need ownership. It needs references

**Comment by codehobbyist06** [comment]
Yes, the C version also stores pointers, but my application is a bit different, in rust I have wrapped objects around those pointers that are being passed around.
Also, I guess I can consider the point of just passing references around and keeping the ownership of the objects in some dump collector.
Thanks a lot for your inputs.

**Comment by jdonszelmann** [comment]
no worries!

**Comment by codehobbyist06** [comment]
Hey one update! I tested your ring buffers push and dequeue APIs against dpdk single packet enqueue and dequeue APIs, the ring buffer performs really good. The earlier comparison was with dpdk bulk enqueue and dequeue.
Following are some stats:
buffer No. of cycles to enqueue No. of cycles to dequeue
dpdk ring 258 291
rust ring 41 68

Looks like if we have the bulk enqueue and dequeue APIs implemented, the performance could be better that dpdk ring buffers.
